### PR TITLE
Change scheduler logic from random to round-robin

### PR DESCRIPTION
**Pull request by gmarek**
To reduce the probability of multiple Pods being schedule on the same node.

cc @hongchaodeng @wojtek-t

### CODE REVIEW

**Comment by gmarek** [comment]
I'm thinking if we shouldn't sort the priority list alphabetically.

**Comment by wojtek-t** [comment]
We definitely should do some sorting, because for maps which we are processing in the meantime, we have exaclty no guarantees...

**Comment by davidopp** [comment]
Yeah I think this only works if `priorityList` and `idx` both remain stable across calls. Is that actually true in practice?

(BTW I didn't understand @wojtek-t 's comment about "maps we are processing in the meantime")

**Comment by wojtek-t** [comment]
>
(BTW I didn't understand @wojtek-t 's comment about "maps we are processing in the meantime")

Sorry - I meant that in the meantime we have maps, and iterating over a map (to generate a slice) is random, so the order is also random.

But yeah - the underlying comment is what @davidopp wrote right above, which is currently not true, but adding sorting will solve this issue.

**Comment by gmarek** [comment]
@wojtek-t @davidopp PTAL

File: plugin/pkg/scheduler/generic_scheduler.go [hunk 1]
```diff
@@ -109,24 +111,16 @@ func (g *genericScheduler) selectHost(priorityList schedulerapi.HostPriorityList
...
-
-	for i, entry := range priorityList {
-		if entry.Score > maxScore {
-			maxScore = entry.Score
-			idx = []int{i}
-		} else if entry.Score == maxScore {
-			idx = append(idx, i)
-		}
-	}
+	lastMaxScoreIndes := sort.Search(len(priorityList), func(i int) bool { return priorityList[i].Score < maxScore })
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/generic_scheduler.go]: to be clear, it's not "lastMaxScoreIndex", but "firstNonMaxScoreIndex"

we should be explicit about this

**Comment by wojtek-t** [comment]
One comment from me - other than that LGTM

**Comment by wojtek-t** [comment]
LGTM

@davidopp - can you please take a look?

File: plugin/pkg/scheduler/api/types.go [hunk 2]
```diff
@@ -165,7 +165,7 @@ func (h HostPriorityList) Less(i, j int) bool {
 	if h[i].Score == h[j].Score {
 		return h[i].Host < h[j].Host
 	}
-	return h[i].Score < h[j].Score
+	return h[i].Score > h[j].Score
```
Comment (davidopp) [inline plugin/pkg/scheduler/api/types.go]: It looks like the more standard way to do this is to leave it as < and then do the sort using 
`sort.Sort(sort.Reverse(priorityList))`

Comment (gmarek) [inline plugin/pkg/scheduler/api/types.go]: Done.

**Comment by gmarek** [comment]
@davidopp PTAL

**Comment by davidopp** [comment]
LGTM

**Comment by davidopp** [comment]
Please apply cherrypick-candidate label if you think we need this in 1.2.

**Comment by gmarek** [comment]
We'll see what impact will it have.

**Comment by gmarek** [comment]
It seems that it prevents spikes in e2e startup latencies in 100 pod/node kubemark runs. We have only 3 runs with this patch now, but if it'll be passing until tomorrow I think we should cherry-pick this change. @wojtek-t @davidopp

**Comment by wojtek-t** [comment]
@gmarek - ok I will check it tomorrow and apply the label if that's the case.

**Comment by timothysc** [comment]
@gmarek could you please /cc the scheduling sigs on these changes.. this effects us too.

**Comment by wojtek-t** [comment]
I've just check our test - and the high density test didn't failed since this PR was merged (which is last 20 runs). We should definitely cherrypick it to 1.2.

**Comment by gmarek** [comment]
@timothysc - sorry, I thought that this is cosmetic enough change, that I don't need to inform anyone about it.

**Comment by jeremyeder** [comment]
Honest question:  is there a way in github to set up certain fields that are required before you can open an issue?

For example, a drop-down picker for "groups to cc" or "tags to add".  Github as a issue reporting/ticketing system is just brutal.

**Comment by eparis** [comment]
Add the v1.2 milestone. Starting later today cherrypick-candidate will be removed from any PR that doesn't set the milestone.

**Comment by bgrant0607** [comment]
@jeremyeder Github just recently added issue and PR templates, if you'd like to investigate how to help us set it up.

**Comment by bgrant0607** [comment]
@gmarek Please write a good release note for this and put it in the PR description.

**Comment by gmarek** [comment]
@bgrant0607 - I'll do this, but this is really just a workaround for broken go 1.4.2 random number generator.

**Comment by eparis** [comment]
This PR was sucessfully cherry picked in PR #22855
please verify that the release-1.2 branch contains these changes as you would expect and contact @eparis if there appear to be problems.

**Pull request merged**